### DISCUSSION TITLE
Curious about intended usage pattern

### DISCUSSION THREAD

**Discussion by unphased** [watchexec/watchexec#675] [Questions & Help] [answered]
So far watchexec has been the only file watcher tool that I've had good success with that is cross platform because it supports FSEvents on macOS and inotify on Linux. Prior to this the tools i've tried (fswatch, entr etc) all just had some sort of critical flaw or flakiness about them that prevented me from being able to standardize. 

However I have come to find by reading some of the manpage files in this project that the environment variable method of pulling out info about watchexec file events is about to be deprecated and there will be json or other type of way to grab this data going forward. 

Unfortunately I'm at a loss as to how to find more detail on this stuff. I suspect that I should be able to continue to rely on the env var communication style for the time being and I have shell code that works reliably with that. 

I just note that sometimes I want to watch and look at some notion of "all" event output but without insane verbosity... So far all i've come up with is `watchexec --print-events 'echo ==='` which works, but I do sometimes see stacks of hundreds of events for a single file being streamed to or something like that. So I was hoping to play around with it if there's a hidden flag or something to enable json output. I would love to just pipe that into jq and do a bit of fiddling to get output streams that can be useful for me to monitor various software's interactions with the filesystem.

**Comment by passcod** [comment]
hmm, that's interesting! there's no direct json-only output like this at the moment, though you could achieve it in two different ways i feel:

1. use the newer json method of getting the changed files. the environment variables getting deprecated (in the long term!) is only because the stdin or file based method is superior and importantly doesn't suffer from the bugs inherent in the "shove it all in the env" way
2. get watchexec to "do nothing" (eg run `echo`) on events and turn on logging in the json format, written to a logfile. Then you can consume that file and filter to the event logs.

**Reply by unphased** [reply]
Yeah I'm referencing where you reference JSON from here: https://github.com/watchexec/watchexec/blob/652b8c9b2fa91858ef15d03fbeff4785b5bec2c9/doc/watchexec.1.md?plain=1#L422

I see this is 8 months old so maybe it was something planned at the time, but not planned any longer. I hope you could elaborate on it more. Maybe this does map to the two methods you just described above, and this sounds consistent so I'm very excited to try them out. Could you point me to documentation detailing them please?

**Reply by passcod** [reply]
i do want to emphathise the longterm-ness of the deprecation of the legacy env "emit events" option!

this project has strong backward compatibility traditions, though not entirely as strong as Rust's... 

what that means is that the option is currently "pre deprecated". in a future major version (probably next year), it will be "officially deprecated", and in practice what that will mean is that it will no longer be the default, but need opt-in. then _at least another six months_ will pass before it's considered for removal, in yet another major version bump.

in other words, you can fully expect that this option will survive until late 2024 at the absolute earliest given other priorities

**Reply by passcod** [reply]
and yeah lemme grab my laptop and i'll put together a couple examples

**Reply by unphased** [reply]
Thanks, I shared my existing code below, and it is def great to know i dont need to port that away from env var usage for a minute. But I'll gladly start early hacking on the better way to go about it. Especially once I get into more advanced behaviors straddling event batches and saving state into files is not the right rabbit hole to go down.

**Comment by unphased** [comment]
I'll share the entire watchexec runner script I'm using in the project of mine that has the most elaborate usage of watchexec right now. 

```
#!/usr/bin/env bash

SELFDIR="${BASH_SOURCE%/*}"

# Simple short circuit checks for a lockout file. This is really only relevant for this metaprogramming project being
# pointed at itself! But it's gonna be a common use case as I'll be dogfooding it frequently during development.
if [ -f "$SELFDIR/watcher.lockout" ]; then
    echo $'\x1b[2m'"Lockout file found, exiting... event: "$'\x1b[22m'"$WATCHEXEC_COMMON_PATH, $WATCHEXEC_WRITTEN_PATH"
    exit
fi

echo $'\x1b[2m'"received a watchexec event at path (showing written files, first one is common path): "$'\x1b[22m'"$WATCHEXEC_COMMON_PATH, $WATCHEXEC_WRITTEN_PATH"

# Get a list of the files that are tracked by git, null-delimited
readarray -d '' git_files < <(git ls-files -z)

PATH_PREFIX="${WATCHEXEC_COMMON_PATH#$(pwd)}"

# Remove leading slash if present and add trailing slash if PATH_PREFIX is not empty
PATH_PREFIX="${PATH_PREFIX:+${PATH_PREFIX#/}/}"

# Assume that WATCHEXEC_WRITTEN_PATH is an environment variable containing 
# colon-separated paths. Split it into an array.
readarray -d ':' -t written_files < <(echo -n "$WATCHEXEC_WRITTEN_PATH")

# Prepend the PATH_PREFIX to each file in the written_files array
for i in "${!written_files[@]}"; do
    written_files[i]="${PATH_PREFIX}${written_files[i]}"
done

TRIGGER_RESTART=0
REBUILD=0

# Modifications on server code triggers full server reload with fancy pgroup kill and pidfile
mapfile -t RESTART_FILES < restart_files.txt

# Modifications on client code triggers rebuild of client code, but not server restart
mapfile -t REBUILD_FILES < rebuild_FE_files.txt

# IGNORE ALL OTHER FILES (one may argue then why all the git ls-files shit but i keep it as an
# example for copying into other projects.)

# enable advanced globbing functionality
shopt -s globstar

# Iterate over each path in the written_files array
for written_file in "${written_files[@]}"; do
    for pattern in "${RESTART_FILES[@]}"; do
        for trigger_file in $pattern; do  # Note the absence of quotes here
            # echo "checking trigger file $trigger_file against $written_file..."
            if [[ "$trigger_file" = "$written_file" ]]; then
                echo $'\x1b[33m'"Server Reinit Triggering file found changed: $written_file"$'\x1b[0m'
                TRIGGER_RESTART=1
                continue 3
            fi
        done
    done
    for rebuild_file in "${REBUILD_FILES[@]}"; do
        if [[ "$rebuild_file" == "$written_file" ]]; then
            echo $'\x1b[35m'"Rebuild Triggering file found changed: $written_file, performing build only"$'\x1b[0m'
            REBUILD=1
            continue 2
        fi
    done
    # Iterate over each file in the git_files array
    for git_file in "${git_files[@]}"; do
        # If the current git_file matches the current written_file, echo it
        # echo "checking $written_file against $git_file..."
        if [[ "$git_file" == "$written_file" ]]; then
            echo $'\x1b[36m'"Some other Git tracked file was found changed, but not monitored by me: $written_file"$'\x1b[0m'
        fi
    done
done

if (( TRIGGER_RESTART )); then
    # kill old process
    set -x
    if [ -f npm_auto_run_pidfile ]; then
        echo sending SIGINT to found pidfile, pid=$(cat npm_auto_run_pidfile)
        PGID=$(ps -o pgid= -p $(cat npm_auto_run_pidfile) | tr -d ' ')
        echo Calling: $'\x1b[31m'"kill -INT -$PGID"$'\x1b[0m' 
        kill -INT -$PGID
        rm npm_auto_run_pidfile
        sleep 0.3
    fi
    npm run start . ~/util/ansi2html.sh &
    # Save pid to pidfile
    echo $! > npm_auto_run_pidfile
    # Some weird construction needed to hide the echo of the set +x being called
    { set +x; :; } 2>/dev/null
    exit
fi

if (( REBUILD )); then
    set -x
    npm run build-frontend-dev-auto
    { set +x; :; } 2>/dev/null
    exit
fi
```

As you can see I do a lot of custom logic here such as enumerating git tracked files, manually maintained file lists for the two different types of trigger behaviors, a little bit of process management, manual path reassembly from events from env vars, etc. 

I would prefer to transition to using javascript to parse events presented in json, I would like to change the interface from being "trigger a file for a batch of events that came in" to "send events in a text stream". This is a more efficient way to go about it. E.g. node.js vm startup time is nontrivial (that's a performance paranoia point, and can be argued, sure)

**Reply by passcod** [reply]
Right, so what you can do is this:

```console
$ watchexec --emit-events-to=json-stdin cat
```

That will produce a stream of json events, one per line. You can literally pipe that into e.g. jq:

```console
$ watchexec --emit-events-to=json-stdin cat | jq
```

for processing, or js or something.

**Reply by passcod** [reply]
For example:

```console
$ touch foo bar

$ watchexec --emit-events-to=json-stdin cat | jq -c

# in another window I did: `touch foo bar`
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"modify","full":"Modify(Metadata(Any))"},{"kind":"path","absolute":"/home/code/watchexec/foo","filetype":"file"}]}
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"modify","full":"Modify(Metadata(Any))"},{"kind":"path","absolute":"/home/code/watchexec/bar","filetype":"file"}]}

# in another window I did: `touch foo baz`
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"modify","full":"Modify(Metadata(Any))"},{"kind":"path","absolute":"/home/code/watchexec/foo","filetype":"file"}]}
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"create","full":"Create(File)"},{"kind":"path","absolute":"/home/code/watchexec/baz","filetype":"file"}]}
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"modify","full":"Modify(Metadata(Any))"},{"kind":"path","absolute":"/home/code/watchexec/baz","filetype":"file"}]}

# in another window I did: `rm foo bar baz`
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"remove","full":"Remove(File)"},{"kind":"path","absolute":"/home/code/watchexec/foo"}]}
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"remove","full":"Remove(File)"},{"kind":"path","absolute":"/home/code/watchexec/bar"}]}
{"tags":[{"kind":"source","source":"filesystem"},{"kind":"fs","simple":"remove","full":"Remove(File)"},{"kind":"path","absolute":"/home/code/watchexec/baz"}]}
```

**Reply by unphased** [reply]
OK. How does this work under the hood, will `cat` be launched once per event batch, and clearly by piping I can get my text event stream like I want... Should I request an option to just send it on stdin without launching any program, just to keep things more simple?

**Reply by passcod** [reply, marked as answer]
Yeah, `cat` is launched once per event batch. Filing a feature request for such an option would be very welcome, it sounds useful and might even be doable soon (currently it's a bit hard under the good to completely disable the process management code, but that's one thing I'm changing soon... though I was looking at it the other way, to have an option to disable filesystem events!).

**Reply by unphased** [reply]
I created a feature request! Sounds good. Thanks for the examples and I appreciate greatly your responsiveness. 

I had to upgrade to 1.23 to get that to work. I always get tripped up thinking the crate is watchexec when it is watchexec-cli. Anyway, good stuff.
